File: workers/research/src/black_book_research/adapters/federal/shared/export_filter.py
"""Filters large federal archive exports before canonical ingestion (BB-046)."""

from __future__ import annotations

import json
from dataclasses import dataclass

from .types import FederalExportFilterPolicy


@dataclass(frozen=True, slots=True)
class ExportFilterResult:
    payload: dict[str, object]
    filtered: bool
    original_bytes: int
    retained_bytes: int
    stripped_keys: tuple[str, ...]


def _estimate_json_bytes(value: object) -> int:
    return len(json.dumps(value, separators=(",", ":")).encode("utf-8"))


def filter_large_export_payload(
    raw: dict[str, object],
    policy: FederalExportFilterPolicy,
) -> ExportFilterResult:
    stripped_keys: list[str] = []
    without_bulk: dict[str, object] = {}

    for key, value in raw.items():
        if key in policy.strip_keys:
            stripped_keys.append(key)
            continue
        without_bulk[key] = value

    original_bytes = _estimate_json_bytes(raw)
    payload = without_bulk
    retained_bytes = _estimate_json_bytes(payload)
    filtered = bool(stripped_keys)

    if retained_bytes > policy.max_payload_bytes:
        payload = {key: without_bulk[key] for key in policy.essential_keys if key in without_bulk}
        retained_bytes = _estimate_json_bytes(payload)
        filtered = True

    return ExportFilterResult(
        payload=payload,
        filtered=filtered,
        original_bytes=original_bytes,
        retained_bytes=retained_bytes,
        stripped_keys=tuple(stripped_keys),
    )

File: packages/domain/src/adapters/chronicling-america/export-filter.ts
/**
 * Filters large Chronicling America exports before candidate stamping.
 * OCR full text and image tile blobs are stripped; only essential metadata is retained.
 */
import type { ChroniclingAmericaExportFilterPolicy } from './types.js';

export type ExportFilterResult = {
  readonly payload: Readonly<Record<string, unknown>>;
  readonly filtered: boolean;
  readonly originalBytes: number;
  readonly retainedBytes: number;
  readonly strippedKeys: readonly string[];
};

function estimateJsonBytes(value: unknown): number {
  return new TextEncoder().encode(JSON.stringify(value)).length;
}

function pickEssentialKeys(
  raw: Readonly<Record<string, unknown>>,
  essentialKeys: readonly string[],
): Record<string, unknown> {
  const retained: Record<string, unknown> = {};
  for (const key of essentialKeys) {
    if (key in raw) {
      retained[key] = raw[key];
    }
  }
  return retained;
}

export function filterLargeExportPayload(
  raw: Readonly<Record<string, unknown>>,
  policy: ChroniclingAmericaExportFilterPolicy,
): ExportFilterResult {
  const strippedKeys: string[] = [];
  const withoutBulk: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(raw)) {
    if (policy.stripKeys.includes(key)) {
      strippedKeys.push(key);
      continue;
    }
    withoutBulk[key] = value;
  }

  const originalBytes = estimateJsonBytes(raw);
  let payload: Record<string, unknown> = withoutBulk;
  let retainedBytes = estimateJsonBytes(payload);
  let filtered = strippedKeys.length > 0;

  if (retainedBytes > policy.maxPayloadBytes) {
    payload = pickEssentialKeys(withoutBulk, policy.essentialKeys);
    retainedBytes = estimateJsonBytes(payload);
    filtered = true;
  }

  return {
    payload,
    filtered,
    originalBytes,
    retainedBytes,
    strippedKeys,
  };
}
